### PR TITLE
examples: new name features

### PR DESCRIPTION
**Pull request by MSusik**
Signed-off-by: Mateusz Susik <mateusz.susik@cern.ch>

### CODE REVIEW

**Comment by MSusik** [comment]
I can imagine that theswe features can be improved, but how?

**Comment by MSusik** [comment]
Before:
```
Number of blocks = 13114
True number of clusters 15575
Number of computed clusters 15811
B^3 F-score (overall) = 0.9800464510382015
B^3 F-score (train) = 0.9867310903532298
B^3 F-score (test) = 0.9796475584448197
```

**Comment by MSusik** [comment]
After:

```
Number of blocks = 13114
True number of clusters 15575
Number of computed clusters 15601
B^3 F-score (overall) = 0.9815810532568245
B^3 F-score (train) = 0.9876010112437174
B^3 F-score (test) = 0.9812041547435574
```
Note that I used old clusters file and not the best sampling strategy for both runs.

File: beard/clustering/blocking.py [hunk 1]
```diff
@@ -50,10 +50,6 @@ def _parallel_fit(fit_, partial_fit_, estimator, verbose, data_queue,
     # by the main process
     status, block, existing_clusterer = data_queue.get()
 
-    if status == 'end':
-        # Special case where there are more processes than blocks
-        result_queue.put((None, None))
-
```
Comment (MSusik) [inline beard/clustering/blocking.py]: These lines were added for a situation where number of processes is bigger than number of blocks. They are obsolete now after I merged 97e930f9113dd416145e9e559b7e863a73eafc86

File: examples/applications/author-disambiguation/distance.py [hunk 2]
```diff
@@ -154,14 +171,14 @@ def _build_distance_estimator(X, y, verbose=0):
             ]), groupby=group_by_signature)),
             ("combiner", CosineSimilarity())
         ])),
-        ("references_similarity", Pipeline([
```
Comment (MSusik) [inline examples/applications/author-disambiguation/distance.py]: Forgot that we reverted it. Will change soon.

File: beard/clustering/blocking_funcs.py [hunk 3]
```diff
@@ -368,14 +368,13 @@ def block_last_name_first_initial(X):
         array is the same as in the ``X`` input parameter.
     """
     def last_name_first_initial(name):
```
Comment (glouppe) [inline beard/clustering/blocking_funcs.py]: The docstring of this function says "The names are normalized before assigning to a block.": we should either respect that or remove that comment from the docstring.

Comment (MSusik) [inline beard/clustering/blocking_funcs.py]: They are still normalized. (Line 371)

File: beard/similarity/pairs.py [hunk 4]
```diff
@@ -464,3 +465,125 @@ def transform(self, X):
             Xt[np.where(denominator == 0)[0]] = 0.
 
         return np.array(Xt).reshape(-1, 1)
+
+
+class CharacterEquality(BaseEstimator, TransformerMixin):
```
Comment (glouppe) [inline beard/similarity/pairs.py]: Seems more generic than character.

File: beard/similarity/pairs.py [hunk 5]
```diff
@@ -464,3 +465,125 @@ def transform(self, X):
...
+        X1, X2 = np.split(X, 2)
+
+        vectorized = np.vectorize(character_equality)
+        n_samples = X1.shape[0]
+
+        val = vectorized(X1, X2)
+        return val.reshape((n_samples, 1))
+
+
+class StringDistance(BaseEstimator, TransformerMixin):
```
Comment (glouppe) [inline beard/similarity/pairs.py]: This class be used to obtain the same as `CharacterEquality` above, by providing the proper similarity function. I would therefore simply remove `CharacterEquality`.

File: beard/utils/names.py [hunk 6]
```diff
@@ -190,6 +190,9 @@ def first_name_initial(name):
     :param name: string
         Name of the author. Usually it should be in the format:
         surnames, first names.
+    :param index: integer
+        Which given name's initial should be returned. 0 for first, 1 for
```
Comment (glouppe) [inline beard/utils/names.py]: s/given/first? (to be consistent with the terminology)

File: beard/utils/names.py [hunk 7]
```diff
@@ -208,3 +214,35 @@ def first_name_initial(name):
             asciified = asciify(split_name[0]).lower().strip()
             return RE_CHARACTERS.findall(asciified)[0]
         return ""
+
+
+@memoize
+def given_name(full_name, index):
+    """Get a specific given name from full name.
```
Comment (glouppe) [inline beard/utils/names.py]: same for here

**Comment by glouppe** [comment]
Other than my comments + what we discussed offline, this is good to go once you have done the changes. Thanks!

File: tests/similarity/test_pairs.py [hunk 8]
```diff
@@ -96,6 +99,22 @@ def test_absolute_difference():
     assert_array_almost_equal(Xt, [[0, 0], [1, 1], [0, 0], [1, 1]])
 
 
+def test_CharacterEquality():
+    """Test for CharacterEquality."""
+    X = np.array(['q', 'q', 'q', '', '', 'q', 'a', '', '', 'q'])
```
Comment (glouppe) [inline tests/similarity/test_pairs.py]: Hmmm, shouldnt you rather test with a 2d array? How can this work with 1d array? (Data is supposed to be "paired" data, with at least 2 features)

File: beard/similarity/pairs.py [hunk 9]
```diff
@@ -464,3 +465,125 @@ def transform(self, X):
...
+        ----------
+        :param X: array-like, shape (n_samples, n_features)
+            Input data.
+
+        Returns
+        -------
+        :returns: Xt array-like, shape (n_samples, 1)
+            The transformed data.
+        """
+        X1, X2 = np.split(X, 2)
```
Comment (glouppe) [inline beard/similarity/pairs.py]: should be `np.split(X, 2, axis=1)` instead

**Comment by glouppe** [comment]
Can I merge this?

**Comment by MSusik** [comment]
Yes, ready.

**Pull request merged**